### PR TITLE
NodeDefEdit: explicit Save

### PR DESCRIPTION
**Pull request by SteRiccio**
When editing a node definition in NodeDefView, the changes will be applied to the DB only when the user presses the button Save.

### CODE REVIEW

File: core/survey/_surveyValidator/nodeDefValidator.js [hunk 1]
```diff
@@ -102,14 +102,14 @@ const validateAdvancedProps = async (survey, nodeDef) => {
   return Validation.newInstance(
     R.all(Validation.isValid, validations),
     R.reject(Validation.isValid, {
-      [NodeDef.propKeys.defaultValues]: validations[0],
-      [NodeDef.propKeys.applicable]: validations[1],
-      [NodeDef.propKeys.validations]: validations[2],
+      [NodeDef.keysPropsAdvanced.defaultValues]: validations[0],
```
Comment (eliask) [inline core/survey/_surveyValidator/nodeDefValidator.js]: Perhaps the preceding lines could deconstruct the `Promise.all` invocation directly, like `const [defaultValue, applicable, validations] = Promise.all(...)`

File: webapp/survey/nodeDefs/actions.js [hunk 2]
```diff
@@ -2,29 +2,58 @@ import axios from 'axios'
...
+
+import { hideAppLoader, showAppLoader } from '@webapp/app/actions'
+import { showNotification } from '@webapp/app/appNotification/actions'
+import { nodeDefEditUpdate } from '@webapp/loggedin/surveyViews/nodeDef/actions'
 
 export const nodeDefCreate = 'survey/nodeDef/create'
-export const nodeDefPropsUpdate = 'survey/nodeDef/props/update'
+export const nodeDefUpdate = 'survey/nodeDef/update'
 export const nodeDefDelete = 'survey/nodeDef/delete'
+export const nodeDefPropsUpdateTemp = 'survey/nodeDef/props/update/temp'
```
Comment (minotogna) [inline webapp/survey/nodeDefs/actions.js]: why temp postfix?

File: webapp/survey/nodeDefs/actions.js [hunk 3]
```diff
@@ -2,29 +2,58 @@ import axios from 'axios'
...
 
+// ==== CREATE
+
+export const createNodeDef = (parent, type, props, history) => async (dispatch, getState) => {
+  const state = getState()
+  const cycle = SurveyState.getSurveyCycleKey(state)
+
+  const nodeDef = {
+    ...NodeDef.newNodeDef(parent, type, cycle, props),
+    [NodeDef.keys.temporary]: true, // Used to dissoc node def on cancel if changes are not persisted
```
Comment (minotogna) [inline webapp/survey/nodeDefs/actions.js]: can you instead assoc it only on save? on nodeDefCreate we shouldn't assoc it

File: webapp/survey/nodeDefs/actions.js [hunk 4]
```diff
@@ -173,25 +188,106 @@ export const putNodeDefLayoutProp = (nodeDef, key, value) => async (dispatch, ge
...
+export const cancelNodeDefEdits = history => async (dispatch, getState) => {
+  const state = getState()
+  const nodeDef = NodeDefState.getNodeDef(state)
+  const nodeDefOriginal = NodeDefState.getNodeDefOriginal(state)
+  const i18n = AppState.getI18n(state)
 
-  const checkFormPageUuid = R.includes(key, [NodeDefLayout.keys.renderType, NodeDefLayout.keys.pageUuid])
-  dispatch(putNodeDefProp(nodeDef, NodeDefLayout.keys.layout, layoutUpdate, false, checkFormPageUuid))
+  if (!NodeDefState.isDirty(state) || confirm(i18n.t('surveyForm.nodeDefEditFormActions.confirmCancel'))) {
+    dispatch({
```
Comment (minotogna) [inline webapp/survey/nodeDefs/actions.js]: I would await this dispatch

File: webapp/survey/nodeDefs/actions.js [hunk 5]
```diff
@@ -173,25 +188,106 @@ export const putNodeDefLayoutProp = (nodeDef, key, value) => async (dispatch, ge
...
+  dispatch(showAppLoader())
+
+  const isNodeDefNew = NodeDef.isTemporary(nodeDef)
+  const nodeDefUpdated = NodeDef.dissocTemporary(nodeDef)
+
+  if (isNodeDefNew) {
+    const {
+      data: { nodeDefsValidation, nodeDefsUpdated },
+    } = await axios.post(`/api/survey/${surveyId}/nodeDef`, nodeDef)
+    _onNodeDefsUpdate(dispatch, nodeDefsUpdated, nodeDefsValidation)
```
Comment (minotogna) [inline webapp/survey/nodeDefs/actions.js]: i would use ```dispatch(_onNodeDefsUpdate(nodeDefsUpdated, nodeDefsValidation))```

File: webapp/survey/nodeDefs/actions.js [hunk 6]
```diff
@@ -40,18 +69,11 @@ const _putNodeDefProps = (nodeDef, props, propsAdvanced) => async (dispatch, get
     propsAdvanced,
   })
 
-  dispatch({ type: nodeDefsValidationUpdate, nodeDefsValidation })
-
-  if (nodeDefsUpdated) {
-    dispatch({ type: nodeDefsUpdate, nodeDefs: nodeDefsUpdated })
-  }
+  _onNodeDefsUpdate(dispatch, nodeDefsUpdated, nodeDefsValidation)
```
Comment (minotogna) [inline webapp/survey/nodeDefs/actions.js]: I would use ```dispatch(_onNodeDefsUpdate(nodeDefsUpdated, nodeDefsValidation)```

File: webapp/survey/nodeDefs/reducer.js [hunk 7]
```diff
@@ -23,12 +30,18 @@ const actionHandlers = {
...
+  [nodeDefUpdate]: (state, { nodeDef }) => NodeDefsState.assocNodeDef(nodeDef)(state),
+
   [nodeDefsUpdate]: (state, { nodeDefs }) => NodeDefsState.mergeNodeDefs(nodeDefs)(state),
+
+  [nodeDefPropsUpdateTemp]: (state, { nodeDef }) => NodeDefsState.assocNodeDef(nodeDef)(state),
+
+  [nodeDefPropsTempCancel]: (state, { nodeDef, nodeDefOriginal, isNodeDefNew }) =>
+    isNodeDefNew
+      ? NodeDefsState.dissocNodeDef(nodeDef)(state) // Remove node def from state
+      : NodeDefsState.assocNodeDef(nodeDefOriginal)(state), // Restore original version of node def
```
Comment (minotogna) [inline webapp/survey/nodeDefs/reducer.js]: why do you even change the nodeDef here when updating props? i suspect you don't need this at all

File: server/modules/survey/manager/surveyManager.js [hunk 8]
```diff
@@ -136,7 +136,7 @@ export const fetchSurveyAndNodeDefsBySurveyId = async (
     fetchSurveyById(surveyId, draft, validate, client),
     NodeDefManager.fetchNodeDefsBySurveyId(surveyId, cycle, draft, advanced, includeDeleted, client),
   ])
-  const survey = Survey.assocNodeDefs(nodeDefs)(surveyDb)
+  const survey = R.pipe(Survey.assocNodeDefs(nodeDefs), Survey.buildAndAssocDependencyGraph)(surveyDb)
```
Comment (minotogna) [inline server/modules/survey/manager/surveyManager.js]: why do you always buildAndAssocDependencyGraph when fetching survey and nodeDefs?

**Comment by eliask** [comment]
Is the general idea here that in the "happy path" the changes still go through, and they can be resubmitted by the client-side saved state? (my understanding, I could be wrong)

I guess that's the least invasive change, but leaves open the possibility of accidentally persisting some changes and leaving them there (or closing a tab? etc.). But that's a tradeoff to make (or not).

File: core/survey/_surveyValidator/nodeDefValidator.js (hunk 1 above)
Comment (SteRiccio) [reply core/survey/_surveyValidator/nodeDefValidator.js]: yes, good catch

File: core/i18n/resources/en.js [hunk 9]
```diff
@@ -407,6 +408,7 @@ export const enTranslation = {
     nodeDefEditFormActions: {
       columns: 'Columns',
       confirmDelete: 'Are you sure you want to permanently delete this node definition?\n\n$t(common.cantUndoWarning)',
+      confirmCancel: 'There are unsaved changes. Do you want to cancel them?',
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply core/i18n/resources/en.js]: ok, as you wish

File: core/survey/nodeDef.js [hunk 10]
```diff
@@ -69,6 +74,9 @@ export const newNodeDef = (nodeDefParent, type, cycle, props, analysis = false)
     ...props,
     [propKeys.cycles]: [cycle],
   },
+  [keys.propsAdvanced]: {
+    ...propsAdvanced,
+  },
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply core/survey/nodeDef.js]: that one creates a shallow copy of the propsAdvanced parameter, I think it's better to keep it.

File: webapp/loggedin/surveyViews/nodeDef/nodeDefState.js [hunk 11]
```diff
@@ -0,0 +1,57 @@
...
+export const getPropsUpdated = getStateProp(keys.propsUpdated, {})
+export const getPropsAdvancedUpdated = getStateProp(keys.propsAdvancedUpdated, {})
+
+// ===== UTILS
+/**
+ * Returns true if nodeDef and nodeDefOriginal are not equals
+ */
+export const isDirty = state => {
+  const nodeDef = getNodeDef(state)
+  return nodeDef && (NodeDef.isTemporary(nodeDef) || !R.equals(nodeDef, getNodeDefOriginal(state)))
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefState.js]: there can be cases where an update reverts a previous update. E.g. update the node def name to "name1" then back to "name": you won't have an empty props object but the node def is the same as the original one.
NodeDefOriginal is used to facilitate the cancel (just assoc the nodeDefOriginal to the survey node defs).

File: webapp/loggedin/surveyViews/nodeDef/nodeDefView.scss [hunk 12]
```diff
@@ -23,8 +23,23 @@
...
 
-  .btn-close {
-    justify-self: center;
+  .button-bar {
+    display: flex;
     align-self: center;
+    justify-content: center;
+  }
+
+  .btn-save {
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefView.scss]: good idea, I'll do it

File: webapp/loggedin/surveyViews/nodeDef/nodeDefView.scss [hunk 13]
```diff
@@ -23,8 +23,23 @@
     grid-template-columns: .2fr .7fr;
   }
 
-  .btn-close {
-    justify-self: center;
+  .button-bar {
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefView.scss]: ok

File: webapp/loggedin/surveyViews/nodeDef/nodeDefView.scss [hunk 14]
```diff
@@ -23,8 +23,23 @@
...
+    text-transform: uppercase;
+
+    &:hover {
+      background-color: lighten($blue, 15%);
+    }
+  }
+
+  .btn-back {
+    margin-right: 3em;
   }
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefView.scss]: okay

File: webapp/loggedin/surveyViews/nodeDef/nodeDefView.js [hunk 15]
```diff
@@ -82,29 +93,38 @@ const NodeDefEditView = props => {
...
+              cancelNodeDefEdits(history)
+            }}
+          >
+            {i18n.t(isDirty ? 'common.cancel' : 'common.back')}
+          </button>
+          <button
+            className="btn btn-save"
+            onClick={() => {
+              saveNodeDefEdits()
+            }}
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefView.js]: absolutely

File: webapp/loggedin/surveyViews/nodeDef/nodeDefView.js [hunk 16]
```diff
@@ -82,29 +93,38 @@ const NodeDefEditView = props => {
...
-          }}
-          aria-disabled={StringUtils.isBlank(NodeDef.getName(nodeDef))}
-        >
-          {i18n.t('common.done')}
-        </button>
+        <div className="button-bar">
+          <button
+            className="btn btn-back"
+            onClick={() => {
+              cancelNodeDefEdits(history)
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefView.js]: right

File: webapp/survey/nodeDefs/actions.js (hunk 2 above)
Comment (SteRiccio) [reply webapp/survey/nodeDefs/actions.js]: I wanted to give enphasys to the fact that the props are not updated immediately to the db, but I can remove it.

File: webapp/survey/nodeDefs/actions.js (hunk 4 above)
Comment (SteRiccio) [reply webapp/survey/nodeDefs/actions.js]: it's better to do it before going back one page, right

File: webapp/survey/nodeDefs/actions.js (hunk 5 above)
Comment (SteRiccio) [reply webapp/survey/nodeDefs/actions.js]: it's an internal method, but just to make the calls consistent I'll change it.

File: webapp/survey/nodeDefs/reducer.js (hunk 7 above)
Comment (SteRiccio) [reply webapp/survey/nodeDefs/reducer.js]: survey node defs are used in several places, they must be up to date with the just applied changes, for example for the expression editor.

File: server/modules/survey/manager/surveyManager.js (hunk 8 above)
Comment (SteRiccio) [reply server/modules/survey/manager/surveyManager.js]: right, it should be done only when "validate" is true

File: webapp/loggedin/surveyViews/nodeDef/nodeDefView.js [hunk 17]
```diff
@@ -82,29 +93,38 @@ const NodeDefEditView = props => {
...
-          className="btn btn-close"
-          onClick={() => {
-            history.goBack()
-            setNodeDefUuidForEdit(null)
-          }}
-          aria-disabled={StringUtils.isBlank(NodeDef.getName(nodeDef))}
-        >
-          {i18n.t('common.done')}
-        </button>
+        <div className="button-bar">
```
[the remark this answers is not in the document]
Comment (SteRiccio) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefView.js]: ok

File: webapp/survey/nodeDefs/actions.js (hunk 6 above)
Comment (SteRiccio) [reply webapp/survey/nodeDefs/actions.js]: ok

File: core/survey/nodeDef.js (hunk 10 above)
[the remark this answers is not in the document]
Comment (minotogna) [reply core/survey/nodeDef.js]: ok.. as you wish

File: webapp/loggedin/surveyViews/nodeDef/nodeDefState.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (minotogna) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefState.js]: but the nodeDefOriginal is already in the state, you can get it through the survey state. why keeping a reference here as well?

File: webapp/loggedin/surveyViews/nodeDef/nodeDefView.js (hunk 17 above)
[the remark this answers is not in the document]
Comment (minotogna) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefView.js]: as agreed, let's just have a common css class

File: webapp/style/main.scss [hunk 18]
```diff
@@ -96,5 +96,33 @@ body {
...
+.button-bar {
+  display: flex;
+  align-self: center;
+  align-items: center;
+  justify-content: center;
+
+  button {
+    margin: 0 .5rem;
+  }
+}
```
Comment (minotogna) [inline webapp/style/main.scss]: I would add these styles where the other button styles are defined (formBase.scss)

File: webapp/survey/nodeDefs/reducer.js (hunk 7 above)
Comment (minotogna) [reply webapp/survey/nodeDefs/reducer.js]: that's true. good

File: webapp/loggedin/surveyViews/nodeDef/nodeDefState.js (hunk 11 above)
[the remark this answers is not in the document]
Comment (minotogna) [reply webapp/loggedin/surveyViews/nodeDef/nodeDefState.js]: no never mind... i think you're right

File: webapp/style/main.scss (hunk 18 above)
Comment (SteRiccio) [reply webapp/style/main.scss]: ok

**Pull request merged**